fix: Cannot Create Multiple Network Prefab Overrides That Target Same Network Prefab as Override [MTT-7399]

NoelStephensUnity opened the pull request:
This resolves the issue where you could not have multiple source network prefab overrides targeting the same network prefab as their override.

[MTT-7399](https://jira.unity3d.com/browse/MTT-7399)

_** Includes additional spawning sample and integration test specific assets (required for integration test)_

Discovered when supporting user requests about NGO documentation in [Isssue-1053](https://github.com/Unity-Technologies/com.unity.multiplayer.docs/issues/1053)

## Addeded/Changed Functionality Update
- Added `NetworkSpawnManafer.InstantiateAndSpawn` that simplifies the spawning of a prefab asset without having to worry about instantiating or making sure to get the overridden version of the prefab when spawning.
  - The included "SimpleSpawn" sample demonstrates how this method simplifies the instantiate and spawn process.
- This update includes automatic linking of in-scene placed `NetworkObject`s to their source prefab asset's `GlobalObjetIdHash` which removes the high user pain associated with having to create a prefab override per in-scene placed network prefab instance within the `NetworkPrefabList`. 
  - _This is specifically useful for when scene management is disabled as it removes the requirement to create a prefab override per in-scene placed `NetworkObject`._ 
  - _This also assures that a prefab with a prefab with override can still be used as an in-scene placed `NetworkObject` without the client spawning the override when scene management is disabled while also still being able to dynamically spawn the same prefab and still yielding the desired override target prefab._
- Increased the default SpawnTimeout to 10s due several user reported issues with messages being dropped prior to a client finishing synchronization.

## Changelog
- Added: `NetworkObject.InstantiateAndSpawn` and `NetworkSpawnManager.InstantiateAndSpawn` methods to simplify prefab spawning by assuring that the prefab is valid and applies any override prior to instantiating the `GameObject` and spawning the `NetworkObject` instance.
- Fixed: Issue where  you could not have multiple source network prefab overrides targeting the same network prefab as their override.
- Changed: `NetworkPrefabs.OverrideToNetworkPrefab` dictionary is no longer used/populated due to it ending up being related to a regression bug and not allowing more than one override to be assigned to a network prefab asset.
- Changed: In-scene placed `NetworkObject`s now store their source network prefab asset's `GlobalObjectIdHash` internally that is used, when scene management is disabled, by clients to spawn the correct prefab even if the `NetworkPrefab` entry has an override. This does not impact dynamically spawning the same prefab which will yield the override on both host and client.
- Changed: In-scene placed `NetworkObject`s no longer require a `NetworkPrefab` entry with `GlobalObjectIdHash` override in order for clients to properly synchronize.
- Changed: In-scene placed `NetworkObject`s now set their `IsSceneObject` value when generating their `GlobalObjectIdHash` value.
- Changed: Increased the default `NetworkConfig.SpawnTimeout` value from 1.0s to 10.0s.

## Testing and Documentation

- Includes Test Project Sample/Manual Test: "SimpleSpawn"
- Includes integration test: `PrefabExtendedTests.TestPrefabsSpawning`
- Documentation Updates: [PR-1142](https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/1142)

NoelStephensUnity committed:
fix

WIP fix for prefab overrides

NoelStephensUnity committed:
update

Removing restriction on PrefabIdHash.
No longer using OverrideToNetworkPrefab.

NoelStephensUnity committed:
update

Removing remains of OverrideToNetworkPrefab

NoelStephensUnity committed:
fix: validation

Putting back the NetworkPrefabs.OverrideToNetworkPrefab to avoid validation issues.

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Update NetworkObject.cs

Removing commented out code.

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
update

Adding a new method: NetworkSpawnManager.InstantiateAndSpawn that handles several steps for users. Basically passing in a prefab NetworkObject will instantiate, spawn, and then return the spawned prefab instance. This takes into account prefab overrides.

Added a manual test "SimpleSpawn" to manually validate the updates.

NoelStephensUnity committed:
style

spelling in xml API
minor rewording

NoelStephensUnity committed:
style

expanded upon the GetNetworkObjectToSpawn  description

NoelStephensUnity committed:
style

fixing namespace ordinal ordering

NoelStephensUnity committed:
update

Added some additional improvements to the simple spawn sample/manual test that also provides an example of the legacy/alternate way to spawn prefabs along with unique prefabs that are named appropriately for each scenario while also setting the color of the NetworkObjectId label to the assigned owner/client assigned color.

NoelStephensUnity committed:
test and update

This update includes automatic linking of in-scene placed NetworkObjects to their source prefab asset. This is specifically useful for when scene management is disabled as it removes the requirement to create a prefab override per in-scene placed NetworkObject. This also assures that a prefab with an override can still be used as an in-scene placed NetworkObject without the client spawning the overide when scene management is disabled while also still spawing the override of the same prefab when dynamically spawned.

This also includes an integration test to validate both with and without scene management enabled that clients synchronize to the proper prefab instance.

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
test update and fix

Reverting the NetworkPrefabs change and just using the SourcePrefabToOverride.
Renamed InScenePlacedSourceGlobalObjectId to InScenePlacedSourceGlobalObjectIdHash to avoid confusion.
Updated the PrefabExtendedTests and test assets to include the scenario that is not handled when scnee management is disabled (i.e. in-scene defined NetworkObjects).

NoelStephensUnity committed:
style

null check standard adjustment

NoelStephensUnity committed:
update

Adding logic to only spawn override (as a default) when running as a host.

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
update

Repopulating the override dictionary to avoid breaking any user's project that utilizes it.
Added NetworkObject.InstantiateAndSpawn method.

NoelStephensUnity committed:
update

Adding change log entries.

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value

NoelStephensUnity committed:
update

Updating how errors and warning messages are handled.

NoelStephensUnity committed:
test

Validating that all error messages occur when expected.

NoelStephensUnity committed:
style

removing CR and adding an s to the end of TestsInstantiateAndSpawnErrors

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefabs.cs:
```diff
@@ -297,12 +300,21 @@ private bool AddPrefabRegistration(NetworkPrefab networkPrefab)
+            if (OverrideToNetworkPrefab.ContainsKey(target))
+            {
+                // If so, just return false and don't attempt to add another.
+                return false;
```
Weren't we going to leave this check out?

NoelStephensUnity replied:
😞 
Yes... it is in the wrong location... need to write another test.

NoelStephensUnity replied:
Ok... check the update... that is where that check should be.

NoelStephensUnity committed:
fix

Putting the OverrideToNetworkPrefab check in the right location.

netcode-ci-service committed:
Merge develop into fix/prefab-override-hash-value

netcode-ci-service committed:
Merge develop into fix/prefab-override-hash-value

NoelStephensUnity committed:
update

Adding back the OverrideToNetworkPrefab to continue to support the legacy manual instantiate and spawn usage pattern.

Updated comments to reflect this and to point users to use InstantiateAndSpawn methods.

NoelStephensUnity committed:
update

Adding a static version of InstantiateAndSpawn to NetworkObject.

NoelStephensUnity committed:
Test

Adding a test pass for validating that the legacy instantiate and spawn usage pattern still works.

NoelStephensUnity committed:
Merge branch 'develop' into fix/prefab-override-hash-value
